### PR TITLE
BallManager God Class

### PR DESCRIPTION
**Pull request by iiKoe**
This pull removes the BallManager as a God Class. The idea of the BallManager (at least where it turned into) has become the GameManager. The new BallManager now handles the ball management. The ball management is further divided up between helper classes. 
The Classes (listed as an "use" hierarchy):
- GameManager
 - BallManager
    - BallsMoving
    - BallsStatic
    - BallsCannon
    - BallsPop
    - BallSnap
    - BallBounce

### CODE REVIEW

File: core/src/com/group66/game/cannon/BallManager.java [hunk 1]
```diff
@@ -1,662 +1,311 @@
...
-    public boolean isGameOver() {
-        for (Ball b : ballStaticList) {
-            // TODO fix the != check
-            if (b.getY() - Config.BALL_DIAM <= Config.BORDER_SIZE_BOT && b.getY() != 0) {
-                return true;
-            }
-        }
-        return false;
+    public boolean hitsBotom() {
+        return ballsStaticManager.hitsBotom();
```
Comment (ALobker) [inline core/src/com/group66/game/cannon/BallManager.java]: bottom

File: core/src/com/group66/game/cannon/GameManager.java [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+        BustaMove.getGameInstance().log(MessageType.Info, "Moved the roof a row down");
+    }
+
+    /**
+     * Checks if it's game over.
+     *
+     * @return true, if is game over
+     */
+    public boolean isGameOver() {
+        return ballManager.hitsBotom();
```
Comment (ALobker) [inline core/src/com/group66/game/cannon/GameManager.java]: bottom

File: core/src/com/group66/game/cannon/GameManager.java [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+                Config.SCORE_OFFSET);
+    }
+
+    /**
+     * Shift clone.
+     *
+     * @param other the other
+     */
+    public void shiftClone(GameManager other) {
+        //TODO change to iterator
```
Comment (ALobker) [inline core/src/com/group66/game/cannon/GameManager.java]: Has this been fixed or can this be fixed?

File: core/src/com/group66/game/cannon/GameManager.java [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    public void ballCheckRoof(Ball ball) {
+        if (ball.getTopHitbox().overlaps(roofHitbox)) {
+            System.out.println("Attach ball to top");
+            ball.setSpeed(0);
+            BallSnap.snapBallToRoof(ball, roofHitbox.y + ROOF_OFFSET, this.isSplit, this.segmentOffset);
+            
+            ballManager.addBallDeadlist(ball);
+            ballManager.addStaticBallToBeAdded(ball);
+            //ballDeadList.add(ball);
+            //ballToBeAdded.add(ball);
```
Comment (ALobker) [inline core/src/com/group66/game/cannon/GameManager.java]: Remove commented code that doesn't work anyways

File: core/src/com/group66/game/cannon/GameManager.java [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        /* Start counting time*/
+        timeKeeper.universalTimeCounter(delta);
+        
+        /* Check if a ball hasn't been shot in the past 10 seconds */
+        timeKeeper.didHeShoot();
+        
+        /* Check shooting balls */
+        // NB. Currently only 1 ball can be shot at a time in the game nevertheless the
+        // current BallList implementation is kept for versatility and to be future proof
+        // TODO change to iterator
```
Comment (ALobker) [inline core/src/com/group66/game/cannon/GameManager.java]: Eclipse can create a list of all TODOs in the code. It's in: Window -> Show View -> Tasks

File: core/src/com/group66/game/cannon/BallManager.java (hunk 1 above)
Comment (iiKoe) [inline core/src/com/group66/game/cannon/BallManager.java]: Spelling is hard

File: core/src/com/group66/game/cannon/BallSnap.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+     *
+     * @param ball the ball
+     * @param hitb the hitball
+     * @param isSplit the is split
+     * @param segmentOffset the segment offset
+     */
+    public static void snapBallToGrid(Ball ball, Ball hitb, boolean isSplit, int segmentOffset) {
+        int xoffset = Config.SINGLE_PLAYER_OFFSET;
+        if (isSplit) {
+            xoffset = Config.SEGMENT_OFFSET * segmentOffset;
```
Comment (iiKoe) [inline core/src/com/group66/game/cannon/BallSnap.java]: What do you mean with this exactly?

File: core/src/com/group66/game/cannon/GameManager.java (hunk 2 above)
Comment (iiKoe) [inline core/src/com/group66/game/cannon/GameManager.java]: Spelling is still difficult

File: core/src/com/group66/game/cannon/GameManager.java (hunk 3 above)
Comment (iiKoe) [inline core/src/com/group66/game/cannon/GameManager.java]: Would be nice to do this in the future, as now you actually have access to the list. Would be better to just give an iteration IMO. But there was no time to do this now.

File: core/src/com/group66/game/cannon/GameManager.java (hunk 5 above)
Comment (iiKoe) [inline core/src/com/group66/game/cannon/GameManager.java]: Nice to know, did you mean this in relation to this piece of code? Or just as a general remark.

File: core/src/com/group66/game/cannon/BallSnap.java (hunk 6 above)
Comment (ALobker) [inline core/src/com/group66/game/cannon/BallSnap.java]: xoffset = Config.SEGMENT_OFFSET * segmentOffset + Config.border_size_sides;

File: core/src/com/group66/game/cannon/GameManager.java (hunk 5 above)
Comment (ALobker) [inline core/src/com/group66/game/cannon/GameManager.java]: General remark

**Pull request merged**